### PR TITLE
NIP-84: Highlights

### PR DESCRIPTION
**Pull request by pablof7z**
This NIP introduces `kind:9802` Highlight event as implemented by https://highlighter.com and https://habla.news

`kind:9802` events are highlights of nostr events' `content` (e.g., `kind:1` or `kind:30023`) or URLs, signaling content of importance to a user.

NIP text: https://github.com/pablof7z/nips/blob/highlights/84.md

@fiatjaf @verbiricha

### CODE REVIEW

File: 84.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+Highlights of nostr-natives events SHOULD use the range index of the content as-is
+(e.g. NIP-23 articles include the markdown instead of computing the index from the rendered markdown).
+
+```
+[ "range", 3000, 3042 ] # highlight begins at index position 3000 of the tagged event's `.content`
+[ "range", 42, 84, "context" ] # highlight begins at index position 42 of the `context` tag's value
+```
+
+#### Non-text-based nostr events's highlights
```
Comment (staab) [inline 84.md:64]: events'

File: 84.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+```json
+[ "p", "...", "author" ],
+[ "p", "...", "author" ],
+[ "p", "...", "editor" ],
+```
+
+### Context
+Clients MAY include a `context` tag, useful when the highlight is a subset of a paragraph and displaying the
+surrounding content might be beneficial to give context to the higlight.
```
Comment (staab) [inline 84.md:37]: Maybe clarify that this is a larger text excerpt of the original content, that wasn't immediately obvious to me. Or just refer to the example below.

File: 84.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+When creating a highlight from a URL the range should be expressed over
+the extracted plain text of the formatted content (e.g. rendered HTML instead of including the HTML markup);
+this helps make finding the correct indexes easier on websites with markup variations on each render.
+
+e.g. `<html><body>hello, world</body></html>`
+
+Tagging `hello, world` would result in using a range tag like `["range", 0, 12 ]`.
+
+Text extraction (i.e. translation from non-plain text medium like HTML or PDF) is highlighy subjective and the value
+of the range should be carefully interpreted by the different clients that support this.
```
Comment (staab) [inline 84.md:80]: Have you found this to be fairly reliable in practice?

**Review by staab — APPROVED** [review]
Do we already have a NIP for bookmarks (other than `e` lists)? This seems like a more granular form of the same thing. Fully support it though.

File: 84.md (hunk 1 above)
Comment (pablof7z) [reply 84.md:64]: damn it! 😂 thanks

File: 84.md (hunk 2 above)
Comment (pablof7z) [reply 84.md:37]: Oh, yeah, good point

File: 84.md (hunk 3 above)
Comment (pablof7z) [reply 84.md:80]: Yeah, there are a handful of libraries that do this text extraction from the DOM, and they seem to be fairly consistent. The index in this case could be thought of more as an approximate value.

File: 84.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+```
+[ "range", 3000, 3042 ] # highlight begins at index position 3000 of the tagged event's `.content`
+[ "range", 42, 84, "context" ] # highlight begins at index position 42 of the `context` tag's value
+```
+
+#### Non-text-based nostr events's highlights
+
+A `kind:9802` event that tags a NIP-94 event which includes a video or audio file can use ranges to
+indicate the start/end time position in seconds.
```
Comment (verbiricha) [inline 84.md:67]: Should we include a tag that marks the highlight as audio/video? It seems to me that audio/video refs could be `r` (youtube link), `a` (replaceable event representing a track) or `e` (NIP-94 audio/video file).

File: 84.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+When creating a highlight from a URL the range should be expressed over
+the extracted plain text of the formatted content (e.g. rendered HTML instead of including the HTML markup);
+this helps make finding the correct indexes easier on websites with markup variations on each render.
+
+e.g. `<html><body>hello, world</body></html>`
+
+Tagging `hello, world` would result in using a range tag like `["range", 0, 12 ]`.
+
+Text extraction (i.e. translation from non-plain text medium like HTML or PDF) is highlighy subjective and the value
```
Comment (verbiricha) [inline 84.md:79]:
```suggestion
Text extraction (i.e. translation from non-plain text medium like HTML or PDF) is highly subjective and the value
```

File: 84.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+Additionally a range with `context` as the third value of the tag MAY be added to indicate the begin/finish indexes
+of the highlight within the included `context` tag.
+
+```
+[ "range", <start-index>, <end-index>, "context" ]
+```
+
+#### Text-based nostr events' highlights
+
+Highlights of nostr-natives events SHOULD use the range index of the content as-is
```
Comment (verbiricha) [inline 84.md:56]:
```suggestion
Highlights of nostr events SHOULD use the range index of the content as-is
```

File: 84.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+(e.g. NIP-23 articles include the markdown instead of computing the index from the rendered markdown).
```
Comment (verbiricha) [inline 84.md:57]: Need to update Habla to use underlying markdown instead of rendered text.

File: 84.md (hunk 1 above)
Comment (fiatjaf) [reply 84.md:64]: You missed a chance to use the most amazing feature of the English language.

**Comment by staab** [comment]
Next release of Coracle will have support for rendering highlights.

File: 84.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+### Attribution
+Clients MAY include one or more `p` tags, tagging the original authors of the material being highlighted; this is particularly
+useful when highlighting non-nostr content for which the client might be able to get a nostr pubkey somehow
+(e.g. prompting the user or reading a `<meta name="nostr:pubkey" />` tag on the document). A role MAY be included as the
+last value of the tag.
+
+```json
+[ "p", "...", "author" ],
+[ "p", "...", "author" ],
+[ "p", "...", "editor" ],
```
Comment (arthurfranca) [inline 84.md:32]:
```suggestion
[ "p", "...", "<recommended relay URL>", "author" ],
[ "p", "...", "<recommended relay URL>", "author" ],
[ "p", "...", "<recommended relay URL>", "editor" ],
```

File: 84.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+Events SHOULD tag the source of the highlight, whether nostr-native or not.
+`a` or `e` tags should be used for nostr events and `r` tags for URLs.
+
+When tagging a URL, clients generating these events SHOULD do a best effort of cleaning the URL from trackers
+or obvious non-useful information from the query string.
+
+### Attribution
+Clients MAY include one or more `p` tags, tagging the original authors of the material being highlighted; this is particularly
+useful when highlighting non-nostr content for which the client might be able to get a nostr pubkey somehow
+(e.g. prompting the user or reading a `<meta name="nostr:pubkey" />` tag on the document). A role MAY be included as the
```
Comment (arthurfranca) [inline 84.md:26]:
```suggestion
(e.g. prompting the user or reading a `<meta name="nostr:nprofile1..." />` tag on the document). A role MAY be included as the
```
nprofiles optionally include relays where clients can use as a hint to get profile metadata from

File: 84.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+Text extraction (i.e. translation from non-plain text medium like HTML or PDF) is highlighy subjective and the value
+of the range should be carefully interpreted by the different clients that support this.
+
+```json
+{
+  "created_at": 1682707885,
+  "content": "while allowing creators to simply keep doing what they’re doing. Creators don’t need to be blatant shills for brands",
+  "tags": [
+    [ "r", "https://footstr.com/zapvertise/" ],
+    [ "p", "c48e29f04b482cc01ca1f9ef8c86ef8318c059e0e9353235162f080f26e14c11", "author" ],
```
Comment (arthurfranca) [inline 84.md:88]:
```suggestion
    [ "p", "c48e29f04b482cc01ca1f9ef8c86ef8318c059e0e9353235162f080f26e14c11", "wss://relay.url", "author" ],
```

File: 84.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+The `.content` of these events is the highlighted portion of the text.
+
+`.content` might be empty for highlights of non-text based media (e.g. NIP-94 audio/video).
+
+### References
+Events SHOULD tag the source of the highlight, whether nostr-native or not.
+`a` or `e` tags should be used for nostr events and `r` tags for URLs.
+
+When tagging a URL, clients generating these events SHOULD do a best effort of cleaning the URL from trackers
+or obvious non-useful information from the query string.
```
Comment (arthurfranca) [inline 84.md:21]: Maybe it is a good idea to say that `r` urls should remove trailing slash so to ease querying.
So https://footstr.com/zapvertise/ -> https://footstr.com/zapvertise
And https://example.com/?article=987 -> https://example.com?article=987

**Review by arthurfranca — CHANGES_REQUESTED** [review]

**Comment by fiatjaf** [comment]
Is this a thing? I think I missed this NIP entirely.

**Comment by pablof7z** [comment]
> Is this a thing? I think I missed this NIP entirely.

with support on amethyst, highlighter, habla and soon(TM) coracle 😉

**Comment by fiatjaf** [comment]
Is the NIP number a reference to some historical event?

**Comment by pablof7z** [comment]
> Is the NIP number a reference to some historical event?

haha, no, this one is random and has no special meaning 😅

File: 84.md (hunk 4 above)
Comment (fiatjaf) [reply 84.md:67]: @pablof7z can you answer this?

File: 84.md (hunk 11 above)
Comment (fiatjaf) [reply 84.md:21]: And also normalize the query strings?

File: 84.md (hunk 10 above)
Comment (fiatjaf) [reply 84.md:88]: @pablof7z please ack here.

**Review by fiatjaf — APPROVED** [review]

File: 84.md (hunk 11 above)
Comment (arthurfranca) [reply 84.md:21]: Good idea. Remove port if 80 (http) or 443 (https) only.

**Review by staab — APPROVED** [review]
I implemented support for this in Coracle yesterday, one thing I'd like to see is an alt/summary/comment tag that allows people to leave a note about the highlight (like a quote tweet). I left a note on highlighter about a highlight, but it didn't show up on the event itself.

**Comment by sroertgen** [comment]
Any specific reason why this is not merged yet?
@pablof7z @fiatjaf @verbiricha @arthurfranca

**Comment by pablof7z** [comment]
I need to update some stuff based on comments but totally forgot; I'll do it soon so we can merge

**Comment by pablof7z** [comment]
> I implemented support for this in Coracle yesterday, one thing I'd like to see is an alt/summary/comment tag that allows people to leave a note about the highlight (like a quote tweet). I left a note on highlighter about a highlight, but it didn't show up on the event itself.

I do this with a kind 1 quoting the kind 9802.

**Comment by alltheseas** [comment]
@pablof7z 

Does the spec allow for combination of a quote note (I believe this is 'q' tag x NIP-18) functionality + highlight functionality?

For instance, if I quote a few words from a note, does NIP-84 specify a way to combine quote note functionality that I can display with the highlight, so that I can add my commentary to the highlight?

cc @ericholguin

**Comment by pablof7z** [comment]
quote highlights are kind:1s that `q` tag a `kind:9802` and have a [ "k", "9802" ] tag. I added some examples in nevent1qgs04xzt6ldm9qhs0ctw0t58kf4z57umjzmjg6jywu0seadwtqqc75spz3mhxw309akx7cmpd35x7um58g6n2deh9uqzqh3mxdlhenu4u7ph2je80qkvplrlaysapjm2lp5uf35y9qkgml4vx7gzjg

**Pull request merged**